### PR TITLE
Cron by ressource

### PR DESCRIPTION
**Pull request by Gabriel-Desharnais**
So this modification integrate the cron inside the web app. There is no need to use crontab. Also there is one cron by resource and each of them can be set at different time interval

### CODE REVIEW

**Comment by justb4** [comment]
Thanks @Gabriel-Desharnais, this is a much requested feature from other users also described in issue #132 by @alexandreleroux . I was wondering if this could be done within a Flask webapp, especially when running in multi-process WGSI deployments like `gunicorn`. As it touches the core of GHC and all existing installations please allow us some time to carefully review your PR.
But again: much appreciated.

As a first step, you may have noticed that the Travis build has failed. Mostly this is due to the Flake8 Python Coding Standards check. It is quite strict but easy to fix and test. See the Travis details. And just run `find . -name "*.py" | grep -v docs | grep -v migrations | xargs flake8`.

Also one of the tests failed, unclear why but your can rerun while fixing:

```
# install GHC with DB etc

# Load testdata into DB
python GeoHealthCheck/models.py load tests/data/fixtures.json y

# Run the healthchecks (what is normally run via cron)
python GeoHealthCheck/models.py run
```

**Comment by Gabriel-Desharnais** [comment]
That should do it

**Comment by justb4** [comment]
@Gabriel-Desharnais @tomkralidis Note some [discussion on Gitter on this PR](https://gitter.im/geopython/GeoHealthCheck?at=5a292c3e87680e6230e11b6e). Probably better to proceed here.

File: GeoHealthCheck/app.py [hunk 1]
```diff
@@ -84,6 +94,33 @@ def db_commit():
     return err
 
 
+def start_crons():
+    # Cold start evry cron of evry ressource
+    for resource in Resource.query.all():
+        freq = resource.test_frequency
```
Comment (justb4) [inline GeoHealthCheck/app.py]: Do this via setting a default value for `test_frequency` in Models.py and the migration script. No need to set here.

File: GeoHealthCheck/app.py [hunk 2]
```diff
@@ -84,6 +94,33 @@ def db_commit():
...
+                          minutes=freq,
+                          id=str(resource.identifier))
+
+    # change configuration
+    scheduler.configure(job_defaults={
+        'coalesce': False,
+        'max_instances': 100000
+        })
+
+    scheduler.add_job(flush_runs, 'interval', minutes=1)
```
Comment (justb4) [inline GeoHealthCheck/app.py]: This is quite often, possibly once an hour will do.

File: GeoHealthCheck/config_main.py [hunk 3]
```diff
@@ -37,7 +37,8 @@
 # Replace None with 'your secret key string' in quotes
 SECRET_KEY = None
 
-GHC_RETENTION_DAYS = 30
+# days, seconds, microseconds, milliseconds, minutes, hours, weeks
+GHC_RETENTION_DAYS = (30, 0, 0, 0, 0, 0, 0)
```
Comment (justb4) [inline GeoHealthCheck/config_main.py]: Try to retain the existing type for `GHC_RETENTION_DAYS`, may expand this to a `timedelta` explicitly.

File: GeoHealthCheck/models.py [hunk 4]
```diff
@@ -497,6 +497,62 @@ def db_commit():
...
+            msg = str(err)
+            print('error notifying: %s' % msg)
+    if not __name__ == '__main__':
+        DB.session.remove()
+
+
+# Complete handle of old runs deletion
+def flush_runs():
+    APP = App.get_app()
+    retention_time = timedelta(*APP.config['GHC_RETENTION_DAYS'])
```
Comment (justb4) [inline GeoHealthCheck/models.py]: See earlier comments: may just calculate retention time from days, e.g. `retention_time ` in milliseconds will be something like: 
`GHC_RETENTION_DAYS * 24 * 3600 * 1000`

File: GeoHealthCheck/templates/edit_resource.html [hunk 5]
```diff
@@ -44,6 +44,12 @@ <h1 class="page-header">[{{ _('Edit') }}] <span id="resource_title_h1">{{ resour
                     </select>
                 </td>
             </tr>
+            <tr>
+                <th>Test frenqucy</th>
+                <td>
+                    <input type="number" id="input_resource_frequency" name="resource_frenquency_value" value="{{ resource.test_frequency }}" style="width: 100%;"/>
```
Comment (justb4) [inline GeoHealthCheck/templates/edit_resource.html]: Spelling: `resource_frenquency_value` should be `resource_frequency_value`

File: GeoHealthCheck/templates/edit_resource.html [hunk 6]
```diff
@@ -204,7 +210,10 @@ <h1 class="page-header">[{{ _('Edit') }}] <span id="resource_title_h1">{{ resour
 
         // Collect tags
         var new_tags = $('#resource_tags').val();
-
+        
+        // Collect test_frequency
+        var new_frequency = $('input[name="resource_frenquency_value"]').val();
```
Comment (justb4) [inline GeoHealthCheck/templates/edit_resource.html]: Spelling: `resource_frenquency_value` should be `resource_frequency_value`

File: GeoHealthCheck/templates/resource.html [hunk 7]
```diff
@@ -80,6 +80,12 @@ <h3 class="page-header">{{ _('Monitoring Period') }}: {{ resource.first_run.chec
                     <button class="btn btn-{{ resource.reliability| cssize_reliability }} btn-sm nohover">{{ resource.reliability|round2 }}%</button>
                 </td>
             </tr>
+            <tr>
+              <th>{{ _('Test frenquency (minutes)') }}</th>
```
Comment (justb4) [inline GeoHealthCheck/templates/resource.html]: Spelling: `frenquency` should be `frequency`

File: GeoHealthCheck/models.py [hunk 8]
```diff
@@ -200,6 +199,7 @@ class Resource(DB.Model):
     owner = DB.relationship('User',
                             backref=DB.backref('username2', lazy='dynamic'))
     tags = DB.relationship('Tag', secondary=resource_tags, backref='resource')
+    test_frequency = DB.Column(DB.Integer)
```
Comment (justb4) [inline GeoHealthCheck/models.py]: add `default` 60 mins here

**Comment by justb4** [comment]
Sorry, but closing this PR, as it has been implemented via PR #222. Thanks @Gabriel-Desharnais for the ideas in this PR. The actual implementation in #222 appeared to be more involved as GHC can run multiple processes in cases, e.g. via WSGI. Also attended via #132.

**Pull request closed without merging**